core/Azure.Mcp.Core/src/Services/ProcessExecution/ExternalProcessService.cs

```diff
@@ -3,21 +3,236 @@
+    {
+        var operationTimeout = ValidateTimeout(operationTimeoutSeconds);
+
+        using Process process = CreateProcess(executablePath, arguments, environmentVariables);
+        using ProcessStreamReader stdoutReader = new(process, isErrorStream: false, _logger);
+        using ProcessStreamReader stderrReader = new(process, isErrorStream: true, _logger);
+
+        try
+        {
+            if (!process.Start())
+            {
+                throw new InvalidOperationException($"Failed to start process: {executablePath}");
+            }
+
+            var stdoutTask = stdoutReader.StartReading();
+            var stderrTask = stderrReader.StartReading();
+            var exitTask = process.WaitForExitAsync(cancellationToken);
+
+            var operation = Task.WhenAll(exitTask, stdoutTask, stderrTask);
+
+            try
+            {
+                await operation.WaitAsync(operationTimeout, cancellationToken).ConfigureAwait(false);
+            }
+            catch (TimeoutException)
+            {
+                // Timeout may be thrown either:
+                //   - Case A: before the process had exited, or
+                //   - Case B: after the process had already exited, but before streams were fully drained.
+                throw HandleTimeout(process, operationTimeout, executablePath, arguments);
+            }
+            catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested)
+            {
+                // Cancellation was explicitly requested by the caller (not a timeout).
+                // OCE may be thrown either:
+                //   - Case A: by WaitForExitAsync while the process is still running, or
+                //   - Case B: by WaitAsync after the process has already exited.
+                HandleCancellation(process, executablePath, arguments, oce);
+                // 'throw;' here preserves the original stack trace from where the OCE was first thrown,
+                // inside WaitAsync or WaitForExitAsync not from this catch block.
+                throw;
+            }
+
+            // The earlier await on Task.WhenAll(...) guarantees that stdoutTask and stderrTask have already run to completion.
+            // The .Result simply retrieves their already-computed output without any blocking.
+            var stdout = stdoutTask.Result.TrimEnd();
+            var stderr = stderrTask.Result.TrimEnd();
+
+            // Normal completion: the process has exited, and stdout/stderr have fully drained.
+            return new ProcessResult(
+                process.ExitCode,
+                stdout,
+                stderr,
+                $"{executablePath} {arguments}");
+        }
+        finally
+        {
+            // ProcessStreamReader.Dispose unsubscribes 'DataReceived' event handlers from the Process.
+            // Process.Dispose releases OS resources.
+        }
+    }
+
+    public JsonElement ParseJsonOutput(ProcessResult result)
```

Comment:
The `ParseJsonOutput` method, and the types it uses `ParseError`, `ParseOutput` are existing (they just look new in this PR but were already part of the`ExternalProcessService`.)